routing/index_graph_starter.cpp

```diff
@@ -214,6 +214,38 @@ set<NumMwmId> IndexGraphStarter::GetMwms() const
   return mwms;
 }
 
+bool IndexGraphStarter::DoesRouteCrossNontransit(
+    RoutingResult<Segment, RouteWeight> const & result) const
+{
+  uint32_t nontransitCrossAllowed = 0;
+  auto isRealOrPart = [this](Segment const & segment) {
+    if (!IsFakeSegment(segment))
+      return true;
+    return m_fake.m_fakeToReal.find(segment) != m_fake.m_fakeToReal.end();
+  };
+  auto isTransitAllowed = [this](Segment const & segment) {
+    auto real = segment;
+    bool const convertionResult = ConvertToReal(real);
+    CHECK(convertionResult, ());
+    return m_graph.GetRoadGeometry(real.GetMwmId(), real.GetFeatureId()).IsTransitAllowed();
+  };
+
+  auto const firstRealOrPart = find_if(result.m_path.begin(), result.m_path.end(), isRealOrPart);
+  if (firstRealOrPart != result.m_path.end() && !isTransitAllowed(*firstRealOrPart))
+    ++nontransitCrossAllowed;
+
+  auto const lastRealOrPart = find_if(result.m_path.rbegin(), result.m_path.rend(), isRealOrPart);
+  // If firstRealOrPart and lastRealOrPart point to the same segment increment
+  // nontransitCrossAllowed once
+  if (lastRealOrPart != result.m_path.rend() && (lastRealOrPart.base() - 1) != firstRealOrPart &&
```

Comment:
```c++
&*lastRealOrPart != &*firstRealOrPart
```
?

Reply:
Optional, from my point of view more readable and less error prone.

Reply:
it's better
done